### PR TITLE
GLTFExporter Tolerance Parameter

### PR DESCRIPTION
**Pull request by fragmuffin**
in response to: #138

### CODE REVIEW

**Comment by fragmuffin** [comment]
@KammutierSpule
I remember why I didn't add this in when I originally wrote it! have a look at 5b416ef72f0b506f864de6db992ed52c3c9a557d to see how much code is required to propagate `tolerance` through the `GLTFExporter`... there's way too much.

Secondly: the call to `tess = shape.tessellate(tolerance)` doesn't seem to have any effect on the output!
However, to test that the given tolerance was going through, I set `tolerance=-1`, and the test started to eat all my RAM, which makes sense.

I can commit this change with a `@skip` on the failing test, but it wouldn't have any functional effect.
I think this change is only relevant for the update to `pythonocc`.

**Comment by KammutierSpule** [comment]
Hi @fragmuffin thanks for look on this!
I was manually changing it in the code the tolerance and it has effect for me.
It depends on the size of the shapes but I had less or more triangles generated. It helps also with the speed ( less / more accuracy ) of the generated mesh while doing some intensive computations.

Looking in the commit 5b416ef, I have no programming knowledge with Python,
but, are you sure if there is some bug mistake? some parts you use self.tolerance other "tolerance = cls.default_tolerance", others "tolerance = self.tolerance", and others "tolerance=tolerance", that confuses me which tolerance is which :)

**Comment by fragmuffin** [comment]
@KammutierSpule 

> I was manually changing it in the code the tolerance and it has effect for me

I've only tried cylinders so far (test-case: [`t_cqparts.test_codecs.TestGltf.test_tolerance`](https://github.com/cqparts/cqparts/blob/5b416ef72f0b506f864de6db992ed52c3c9a557d/tests/t_cqparts/test_codecs.py#L345-L368)).
What shape did you use to get a change in the polygon count?

> Looking in the commit 5b416ef [...] are you sure if there is some bug mistake?

No; I think it's correct, the different tolerance variables are used at different scopes:
* `cls.default_tolerance` - when no `tolerance` value is given
* `self.tolerance` - when the `GLTFExporter` instance was given a `tolerance`
* `tolerance=tolerance` - at the lowest scope (scoped to _just_ the function).

It was all done by hand, so there shouldn't be any copy/paste errors in there :+1: 

> I have no programming knowledge with Python

I think you've got more much more than _no_ knowledge. You're not just using open source software, but contributing to it! People with no knowledge can't do that :+1:

**Comment by KammutierSpule** [comment]
> What shape did you use to get a change in the polygon count?
You can try a cut operation between two spheres.
As I understand unities are in millimeters, so if you start with two 10.0 mm radius spheres you can try increase the tolerance until you see differences.
I am using the exporter generated files and open it with Blender. You can also use it to debug triangles or see the polygon count, but the differences by changing tolerance should be noticeable by eye as tolerance increase to great values, it will produce more blocky results.
As you point on documentation, this only affects curve surfaces ( I believe cylinders too )

> I think you've got more much more than no knowledge. You're not just using open source software, but contributing to it! People with no knowledge can't do that 

yeah thanks, I mean python syntax is new for me so I am limited.

**Comment by fragmuffin** [comment]
@KammutierSpule - hmm, I'm still not getting any change in the polygon count.
I've changed the test locally to use the shape you suggested (sphere cut from another sphere).
But the polygon count isn't changing (tested on both py2 and py3)

```python
import cqparts
import cadquery

class SphereCut(cqparts.Part):
    def make(self):
        sphere1 = cadquery.Workplane('XY', origin=(0,0,0)).sphere(10)
        sphere2 = cadquery.Workplane('XY', origin=(10,0,0)).sphere(10)
        return sphere1.cut(sphere2)

from cqparts.display import display
from cqparts.display.material import TEMPLATE
obj = SphereCut(_render=TEMPLATE['wood'])
display(obj)  # opens shape below in browser

# export instead
obj.exporter('gltf')('spherecut.gltf', tolerance=0.05)
```

image

Just to confirm, are we both talking about the `master` branch that uses FreeCAD (ie: not the `feature/freecad-occ` branch)
What version of FreeCAD are you running?

**Comment by fragmuffin** [comment]
ok yeah! that's a dramatic change in the poly-count...
However, I'm just not seeing any change on my system.

**My Environment Versions**
For pushes to `master` I'm mainly using docker environment [`ubuntu-py3`](https://github.com/cqparts/cqparts/blob/master/env/ubuntu-py3).
But i'm seeing the same behaviour (no polygon count change) from all of:
* docker `ubuntu-py3` - freecad `v0.18b0.py36h275e31e_2` python `v3.5.2`
* docker `ubuntu-py2` - freecad `v0.17.13541` python `v2.7.12`
* my host (also ubuntu) which uses freecad `v0.17.13522` python `v2.7.12`
* appveyor test environment (windows-x64) - freecad `v0.17` and python `v3.6`

Then on top of that, `ubuntu-py2`, `ubuntu-py3`, and `appveyor` are _all_ tested on an independent system when these commits go through to this PR...
I'm running out of ideas

wait... OMG...

**HOLD THE PHONE :phone: :no_entry:**

**This** code *works*!

```python
import cqparts
import cadquery

class SphereCut(cqparts.Part):
    def make(self):
        sphere1 = cadquery.Workplane('XY', origin=(0,0,0)).sphere(10)
        sphere2 = cadquery.Workplane('XY', origin=(10,0,0)).sphere(10)
        return sphere1.cut(sphere2)

#obj = SphereCut()  # this was the problem

def sizes(t):
    obj = SphereCut()
    e = obj.exporter('gltf')
    buf = e.part_buffer(obj, tolerance=t)
    mesh = e.part_mesh(obj, tolerance=t)
    return (buf.idx_size // 3, len(mesh[1]))

print(sizes(0.01))
print(sizes(0.10))
```

Output:
```
(14076, 14076)
(2246, 2246)
```

However: if the creation of `obj` is made global (commented out) instead of inside the function, the same instance is exported twice, and the output is:

```
(14076, 14076)
(14076, 14076)
```

However again: if the `obj` is made global, and the tolerance testing is done in reverse:
```python
print(sizes(0.10))  # big number first
print(sizes(0.01))
```

Then we get the correct output:
```
(2246, 2246)
(14076, 14076)
```

I think FreeCAD is caching the mesh somewhere; I think if you ask to tessellate a shape with `0.01`, then ask it again with a larger number, the tessellation engine says "hey, the mesh I have is within that tolerance, here you go"
(note: making the exporter global doesn't change anything, so I don't think `cqparts` is caching anything)

**Comment by fragmuffin** [comment]
@adam-urbanczyk @jmwright @dcowden @justbuchanan @RustyVermeer
Some quick questions for you...

1. Have you seen the freecad tessellation method do this before?
1. Is this something to compensate for in pythonocc?

(details in the [previous comment](https://github.com/cqparts/cqparts/pull/141#issuecomment-450720004))

**TL;DR**: In relation to
https://github.com/cqparts/cqparts/blob/07fd3d4cc3bd0d4cbc987cfbdd0f47daa0a734e9/src/cqparts/codec/gltf.py#L491-L492

> I think FreeCAD is caching the mesh somewhere; I think if you ask to tessellate a shape with [a tolerance] of 0.01, then ask it again with a larger number, the tessellation engine says "hey, the mesh I have is within that tolerance, here you go"

**Comment by fragmuffin** [comment]
@KammutierSpule 

> it would be helpful to have also a tolerance parameter [for `display()`]

I agree :+1:

**Comment by KammutierSpule** [comment]
Nice that you find the cause!
I think it would be a known bug I could leave with, since most of the cases I will use a global tolerance for the output.

**Comment by jmwright** [comment]
> 1. Have you seen the freecad tessellation method do this before?
> 2. Is this something to compensate for in pythonocc?

I have not seen that happen with #1. I can't answer about #2.

**Comment by dcowden** [comment]
@fragmuffin I have never meshed with FreeCAD, successfully or cached ;)  it sure seems weird. if your theory is right that FreeCAD is caching it, then it becomes a non-issue using pythonOCC version of CQ.

**Comment by KammutierSpule** [comment]
I found that just calling
`bbox = part.bounding_box`
will also cause the mesh to be pre-computed ( and so, cached ), so tolerance will not have any effect if it is used after it :/
